src/go/rdctl/cmd/snapshot.go

```diff
@@ -63,38 +63,46 @@ func exitWithJsonOrErrorCondition(e error) error {
 // If the main process is running, stops the backend, calls the
 // passed function, and restarts the backend. If it cannot connect
 // to the main process, just calls the passed function.
-func wrapSnapshotOperation(wrappedFunction cobraFunc) cobraFunc {
-	return func(cmd *cobra.Command, args []string) error {
-		appPaths, err := paths.GetPaths()
-		if err != nil {
-			return fmt.Errorf("failed to get paths: %w", err)
-		}
-		if err := createBackendLock(appPaths.AppHome); err != nil {
-			return err
-		}
-		defer removeBackendLock(appPaths.AppHome)
-		if err := ensureBackendStopped(cmd); err != nil {
-			return err
-		}
-		if err := wrappedFunction(cmd, args); err != nil {
+func wrapSnapshotOperation(cmd *cobra.Command, appPaths paths.Paths, resetOnFailure bool, wrappedFunction cobraFunc) error {
+	if err := createBackendLock(appPaths.AppHome); err != nil {
+		return err
+	}
+	defer removeBackendLock(appPaths.AppHome)
+	if err := ensureBackendStopped(cmd); err != nil {
+		return err
+	}
+	if err := wrappedFunction(); err != nil {
+		if resetOnFailure {
 			factoryreset.DeleteData(appPaths, true)
-			return err
 		}
-		// Note that this does not wait for the backend to be in the
-		// STARTED (or DISABLED if k8s is disabled) state. This allows
-		// removeBackendLock() to be called as a deferred function while
-		// keeping the state of the backend lock file in sync with the
-		// main process backendIsLocked variable.
-		return ensureBackendStarted()
+		return err
```

Comment:
Why are we not restarting the backend here (only when the backend was running before we invoked the wrapped function)?

Reply:
> only when the backend was running before we invoked the wrapped function

Didn't we say that we wanted to try restarting the backend, assuming no error in the snapshot operation, whether we were able to shut it down? See #5674.

We need to look at this function in the context of the two places it is used:

Snapshot creation: when `manager.Create()` returns an error, the backend should be restarted and then the error should be returned.

Snapshot restoration: when `manager.Restore()` returns an error, it should do a factory reset and then return the error.

Reply:
Also, I still think it would be easier (both now and in the future when we make modifications) to do away with `wrapSnapshotOperation` and put the necessary code in `createSnapshot()` and `restoreSnapshot()`.

Reply:
This isn't an actual change in this PR -- you wouldn't see this in `git diff --color-moved`.